### PR TITLE
Rates testing

### PR DESCRIPTION
**Pull request by jamesaimonetti**
Tests tasks a bit too. Makes sure adding rates to multiple ratedecks works as expected.

### CODE REVIEW

File: applications/crossbar/priv/couchdb/views/rates.json [hunk 1]
```diff
@@ -3,10 +3,22 @@
     "language": "javascript",
     "views": {
         "crossbar_listing": {
-            "map": "function(doc) { if (doc.deleted || doc.pvt_type !== 'rate') return; emit(doc._id, {\"name\":doc.rate_name, \"cost\":doc.rate_cost, \"prefix\":doc.prefix, \"description\":doc.description, \"surcharge\":doc.rate_surcharge, \"carrier\":doc.carrier, \"id\":doc._id}); }"
+            "map": [
+                "function(doc) {",
+                "  if ( doc.pvt_deleted || doc.pvt_type !== 'rate') return;",
+                "  emit(doc._id, {",
+                "\"name\":doc.rate_name,",
```
Comment (fenollp) [inline applications/crossbar/priv/couchdb/views/rates.json]: Keys don't need to be quoted. If you really want to quote them you can use `'`.
Formatting: please indent & add a space after `:`

File: applications/crossbar/src/crossbar.hrl [hunk 2]
```diff
@@ -106,6 +106,7 @@
                          ,'cb_services'
                          ,'cb_simple_authz'
                          ,'cb_sms'
+                         ,'cb_tasks'
```
Comment (fenollp) [inline applications/crossbar/src/crossbar.hrl]: `cb_tasks` needs `tasks` to be running. Is this required?

File: applications/tasks/src/kz_task_worker.erl [hunk 3]
```diff
@@ -223,7 +225,7 @@ is_task_successful(MappedRow
                     {'false', Columns, Written, NewIterValue}
             end;
         Fields ->
-            PPFields = kz_binary:join(Fields, $\s),
+            PPFields = kz_binary:join(Fields, <<$\s>>),
```
Comment (fenollp) [inline applications/tasks/src/kz_task_worker.erl]: This does not need to be a binary. Why waste the memory?

File: core/kazoo_documents/src/kzd_rate.erl [hunk 4]
```diff
@@ -49,18 +49,17 @@ from_map(Map) ->
     ensure_id(set_type(Rate)).
 
 -spec ensure_id(doc()) -> doc().
+-spec ensure_id(doc(), api_ne_binary()) -> doc().
 ensure_id(Rate) ->
-    case kz_doc:id(Rate) of
-        'undefined' -> set_rate_id(Rate);
-        _Id -> Rate
-    end.
+    ensure_id(Rate, kz_json:get_ne_binary_value(<<"_id">>, Rate)).
```
Comment (fenollp) [inline core/kazoo_documents/src/kzd_rate.erl]: This is `kz_doc:id/1`

File: doc/blog/understanding_rates.md [hunk 5]
```diff
@@ -5,8 +5,8 @@ System administrators can create multiple ratedecks and assign them to accounts
 
 ## High level overview
 
-1.  Admins can create ratedecks by uploading CSVs to the [Tasks API](/applications/crossbar/doc/tasks.md) and defining the \`ratedeck\_name\`. See the [Rate Task](/applications/tasks/doc/rates.md) docs for how to upload the CSV. This will be covered further down too.
-2.  Admins can then create [Service Plans](/applications/crossbar/doc/service_plans.md) that incorporate the ratedeck(s) available. You can optionally charge for access to these ratedecks as part of the service plan.
+1.  Admins can create ratedecks by uploading CSVs to the [Tasks API](file:///home/james/local/git/2600hz/kazoo/applications/crossbar/doc/tasks.md) and defining the \`ratedeck\_name\`. See the [Rate Task](file:///home/james/local/git/2600hz/kazoo/applications/tasks/doc/rates.md) docs for how to upload the CSV. This will be covered further down too.
```
Comment (fenollp) [inline doc/blog/understanding_rates.md]: Paths.

File: doc/blog/understanding_rates.md [hunk 6]
```diff
@@ -383,7 +385,7 @@ Now that we have our ratedecks, let's create service plans that can be applied t
 
 ### Service Plan format
 
-Take a look at the [Service Plan lifecycle](/core/kazoo_services/doc/lifecycle.md) for details on how service plans work. Basically, we have a category of `ratedeck` and an item with the `ratedeck_id` as the key. You can also optionally include surcharges, activation fees, etc.
+Take a look at the [Service Plan lifecycle](file:///home/james/local/git/2600hz/kazoo/core/kazoo_services/doc/lifecycle.md) for details on how service plans work. Basically, we have a category of `ratedeck` and an item with the `ratedeck_id` as the key. You can also optionally include surcharges, activation fees, etc.
```
Comment (fenollp) [inline doc/blog/understanding_rates.md]: Path.

File: core/kazoo_proper/src/pqc_cb_api.erl [hunk 7]
```diff
@@ -21,7 +21,7 @@
              ]).
 
 -define(API_BASE, "http://localhost:8000/v2").
--define(API_KEY, <<"df52943da7119a3fc901dcadba313c0689d6d883dd26f5fa84abbc44df6c30d1">>).
+-define(API_KEY, <<"d056f9d536914499dd1e8cf63593be4ea333368adf630583a38db30b1877a710">>).
```
Comment (fenollp) [inline core/kazoo_proper/src/pqc_cb_api.erl]: Curious. Why this change?

File: core/kazoo_documents/src/kzd_rate.erl (hunk 4 above)
Comment (jamesaimonetti) [inline core/kazoo_documents/src/kzd_rate.erl]: It's not, `kz_doc:id/1` tries to read "id" and "ID" as well. I wanted to be explicit in finding the "_id".

File: applications/tasks/src/kz_task_worker.erl (hunk 3 above)
Comment (jamesaimonetti) [inline applications/tasks/src/kz_task_worker.erl]: It was either update the spec for join or change the arg here. Wasted memory = 6 words?

File: applications/crossbar/src/crossbar.hrl (hunk 2 above)
Comment (jamesaimonetti) [inline applications/crossbar/src/crossbar.hrl]: I think tasks and cb_tasks should be required at this point (4.2/master) no?

File: core/kazoo_proper/src/pqc_cb_api.erl (hunk 7 above)
Comment (jamesaimonetti) [inline core/kazoo_proper/src/pqc_cb_api.erl]: My master account api key changed

File: applications/tasks/src/kz_task_worker.erl (hunk 3 above)
Comment (fenollp) [inline applications/tasks/src/kz_task_worker.erl]: what's the cost of updating a wrong spec?

File: core/kazoo_documents/src/kzd_rate.erl (hunk 4 above)
Comment (fenollp) [inline core/kazoo_documents/src/kzd_rate.erl]: it reads _id first. If you don't trust kz_doc:id/1 you're welcome to test it.

File: core/kazoo_proper/src/pqc_cb_api.erl (hunk 7 above)
Comment (fenollp) [inline core/kazoo_proper/src/pqc_cb_api.erl]: How about reading it from couch then?

File: applications/crossbar/src/crossbar.hrl (hunk 2 above)
Comment (fenollp) [inline applications/crossbar/src/crossbar.hrl]: That's a @k-anderson question I think

File: applications/crossbar/priv/couchdb/views/rates.json [hunk 8]
```diff
@@ -3,10 +3,22 @@
...
+                "  if ( doc.pvt_deleted || doc.pvt_type !== 'rate') return;",
+                "  emit(doc._id, {",
+                "    'name': doc.rate_name,",
+                "    'cost': doc.rate_cost,",
+                "    'prefix': doc.prefix,",
+                "    'description': doc.description,",
+                "    'surcharge': doc.rate_surcharge,",
+                "    'carrier': doc.carrier,",
+                "    'id': doc._id",
+                "}); }"
```
Comment (fenollp) [inline applications/crossbar/priv/couchdb/views/rates.json]: Sorry didn't see that before: can you drop the `}` on a new line?

File: core/kazoo_documents/src/kzd_rate.erl (hunk 4 above)
Comment (jamesaimonetti) [inline core/kazoo_documents/src/kzd_rate.erl]: Which is fine if _id exists; but if _id is missing but id exists, it won't set _id and you get a 32-char uuid for id.

File: core/kazoo_proper/src/pqc_cb_api.erl (hunk 7 above)
Comment (jamesaimonetti) [inline core/kazoo_proper/src/pqc_cb_api.erl]: that was my thought eventually, esp once this can be a CI-like or make-busy-like suite

File: core/kazoo_documents/src/kzd_rate.erl (hunk 4 above)
Comment (fenollp) [inline core/kazoo_documents/src/kzd_rate.erl]: Okay

File: applications/tasks/src/modules/kt_rates.erl [hunk 9]
```diff
@@ -241,12 +243,22 @@ cleanup(<<"delete">>, State) ->
...
+    lager:debug("importing ~p rate~s into ratedeck '~s'"
+               ,[_C, maybe_plural(_C), RatedeckDb]
+               ),
+
     kz_datamgr:suppress_change_notice(),
-    save_rates(kzd_ratedeck:format_ratedeck_db(Ratedeck), Rates),
+    save_rates(RatedeckDb, Rates),
     import_rates_into_ratedeck(Ratedeck, {0, []}).
 
+-spec maybe_plural(pos_integer()) -> [] | [$s].
```
Comment (fenollp) [inline applications/tasks/src/modules/kt_rates.erl]: Specs can't get char syntax for some reason. Plus: `[] | [something()]` is exactly `[something()]`.
How about just `string()`?

**Comment by jamesaimonetti** [comment]
@fenollp ready
